text/0000-extern-existential-type.md

```diff
@@ -0,0 +1,238 @@
+
+Many of the mechanisms listed in this RFC above are on the verge of stabilization.
+This RFC doesn't want to appear to by tying things up forever, so the design strives to be simple while still being general enough.
+This ought to also be forwards compatible with the more comprehensive solutions as described in the [alternatives](#alternatives) section.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+It's best to understand this feature in terms of regular `existential type`.
+Type checking when *using* `pub extern existential type` works exactly the same way:
+The type is opaque except for its bounds, and no traits can be implemented for it.
+```rust
+pub extern existential type Foo: Baz;
+existential type Bar: Baz;
+// both are used the same way in other modules
+```
+C and C++ programmers too will be familiar with the remote definition aspect of this from those language's "forward declarations" and their use in header files.
+
+On the definition side, since it is explicitly defined (somewhere), there are no inference constraints on items in the same module as the declaration or definition.
+
+One more interesting difference is the scope of where the type is transparent vs opaque: i.e. where can we see the type's definition, or only it's bounds.
+Just as in C where one gets:
+```rust
+struct Foo;
+
+// I know nothing about Foo
+
+struct Foo { int a; };
+
+// Ah now I do
+```
+when the `extern existential type` is in scope, the `existential existential type` becomes transparent and behaves as if the declaration and definition were put together into a normal type alias.
+The definer can decide how one downstream gets to take advantage of it by making the definition public or not.
+```rust
+pub extern existential type alloc::Foo = Bar; // the big reveal
+extern existential type alloc::Foo = Bar; // the tiny reveal
+```
+private allows the item to be used (as some definition is needed), but while no one downstream knows its true definition. like regular `existential type`.
+Public allows downstream to choose between staying agnostic for increased flexibility, or peaking the hind the veil for extra functionality.
+(e.g. maybe it wants to require the global allocator by jemalloc to use some special jemalloc-specific debug output.)
+There are no restrictions on the type of publicity on the definitions compared to other items.
+
+Only one crate in the build plan can define the `pub extern existential type`.
+Unlike the trait system, there are no orphan restrictions that ensure crates can always be composed:
+any crate is free to define the `pub extern existential type`, as long is it isn't used with another that also does, in which case the violation will only be caught when building a crate that depends on both (or if one of the crates depends on the other).
+This is not very nice, but exactly like "lang items" and the annotations that exist for this purpose today,
```

Comment:
This sentence and the rest of this paragraph shouldn't be in the "guide-level explanation"